presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,211 @@
+ * limitations under the License.
+ */
+package io.prestosql.plugin.prometheus;
+
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Files;
+import com.google.inject.ConfigurationException;
+import com.google.inject.spi.Message;
+import io.airlift.json.JsonCodec;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.spi.type.DoubleType;
+import io.prestosql.spi.type.MapType;
+import io.prestosql.spi.type.TimestampType;
+import io.prestosql.spi.type.TypeManager;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import org.apache.http.client.utils.URIBuilder;
+
+import javax.inject.Inject;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static io.prestosql.plugin.prometheus.PrometheusSplitManager.timeUnitsToSeconds;
+import static io.prestosql.spi.StandardErrorCode.NOT_FOUND;
+import static io.prestosql.spi.StandardErrorCode.REMOTE_TASK_ERROR;
+import static io.prestosql.spi.type.TypeSignature.mapType;
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusClient
+{
+    // schema name is fixed to "default"
+    private final Supplier<Map<String, Object>> tableSupplier;
+    protected static PrometheusConnectorConfig config;
+    protected static final String METRICS_ENDPOINT = "/api/v1/label/__name__/values";
+    static List<PrometheusColumn> prometheusColumns;
+    private Optional<TupleDomain<ColumnHandle>> pushedDown = Optional.empty();
+    private static final OkHttpClient httpClient = new OkHttpClient.Builder()
+            .build();
+
+    @Inject
+    public PrometheusClient(PrometheusConnectorConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
+            throws URISyntaxException
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(metricCodec, "metricCodec is null");
+        requireNonNull(typeManager, "typeManager is null");
+
+        long maxQueryRangeDuration = timeUnitsToSeconds(config.getMaxQueryRangeDuration());
+        long queryChunkSizeDuration = timeUnitsToSeconds(config.getQueryChunkSizeDuration());
+        if (maxQueryRangeDuration < queryChunkSizeDuration) {
+            throw new ConfigurationException(ImmutableList.of(new Message("max-query-range-duration must be greater than query-chunk-size-duration")));
```

Comment:
Validation of configuration should happen in config class. Use `@PostConstruct` for validation method.

Reply:
Fixed in aacb66d8

Reply:
Ah but actually, there is this unit test failing now though:
```java
    @Test
    public void testFailOnDurationLessThanQueryChunkConfig()
            throws Exception
    {
        PrometheusConnectorConfig config = new PrometheusConnectorConfig();
        config.setPrometheusURI(new URI("http://doesnotmatter.com"));
        config.setQueryChunkSizeDuration("21d");
        config.setMaxQueryRangeDuration("1d");
        config.setCacheDuration("30s");
        assertThatThrownBy(() -> new PrometheusClient(config, METRIC_CODEC, TYPE_MANAGER))
                .isInstanceOf(ConfigurationException.class)
                .hasMessageContaining("max-query-range-duration must be greater than query-chunk-size-duration");
    }
```
and I'm not immediately sure how to fix it given it's being called via the @PostConstuct annotation.

Reply:
Well I think the answer is just all the `@PostConstruct` annotated method directly in the test. Which is what is done in 6d70e1f7. But leaving this convo unresolved to highlight that choice in case that's not a good approach.